tests/lib/metadata_linux_bridge_test.py

```diff
@@ -0,0 +1,365 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+#
+# This file is part of nmstate
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import metadata
+from libnmstate import state
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceState
+from libnmstate.schema import InterfaceType
+from libnmstate.schema import LinuxBridge
+
+
+BRIDGE0 = "br0"
+BRIDGE1 = "br1"
+IFACE0 = "eth0"
+IFACE1 = "eth1"
+
+
+class TestLinuxBridgeMetadata:
+    def test_creation_with_new_ports(self):
+        desired_state = state.State(
+            {
+                Interface.KEY: [
+                    _create_bridge_state(BRIDGE0, (IFACE0, IFACE1)),
+                    _create_iface_state(IFACE0, InterfaceType.UNKNOWN),
+                    _create_iface_state(IFACE1, InterfaceType.UNKNOWN),
+                ]
+            }
+        )
+        current_state = state.State({})
+        expected_dstate = state.State(
+            {
+                Interface.KEY: [
+                    _create_bridge_state(BRIDGE0, (IFACE0, IFACE1)),
+                    _create_iface_state(
+                        IFACE0,
+                        InterfaceType.UNKNOWN,
+                        extra=_create_metadata_state(BRIDGE0, IFACE0),
+                    ),
+                    _create_iface_state(
+                        IFACE1,
+                        InterfaceType.UNKNOWN,
+                        extra=_create_metadata_state(BRIDGE0, IFACE1),
+                    ),
+                ]
+            }
+        )
+        expected_cstate = state.State(current_state.state)
+
+        metadata.generate_ifaces_metadata(desired_state, current_state)
```

Comment:
This test shows that `metadata.generate_ifaces_metadata()` is badly designed, since it seems to modify its arguments instead of returning the changed data. It should probably be a method of `state.State()` anyhow. Just a note, though.

Reply:
It is mutable indeed. Refactoring this to an immutable version is welcome indeed.
I do not see how it is related to `State`.

Reply:
The metadata is a property of the state - or actually a workaround for passing around a dictionary instead of an object with properties or actually the items of the metadata should be properties of a InterfaceState object, I guess.